packages/cw-snapshot-vector-map/src/lib.rs

```diff
@@ -0,0 +1,409 @@
+        let expiration = expire_in.map(|d| curr_height + d);
+        active.push((next_id, expiration));
+        self.active.save(store, k.clone(), &active, curr_height)?;
+
+        // update next ID
+        self.next_ids.save(store, k.clone(), &(next_id + 1))?;
+
+        Ok(((next_id, expiration), active.len()))
+    }
+
+    /// Removes an item from the vector by ID and returns it, along with the
+    /// total number of items remaining. The block height should be greater than
+    /// or equal to the blocks all previous items were added/removed at.
+    /// Removing from the past will lead to incorrect behavior.
+    pub fn remove(
+        &self,
+        store: &mut dyn Storage,
+        k: &K,
+        id: u64,
+        curr_height: u64,
+    ) -> StdResult<(V, usize)> {
+        // ensure remove operations are performed at or after the last update
+        self.validate_order_and_update(store, k, curr_height)?;
+
+        // get active list for the key
+        let mut active = self.active.may_load(store, k.clone())?.unwrap_or_default();
+
+        // remove item and any expired items
+        active.retain(|(active_id, expiration)| {
+            active_id != &id && expiration.map_or(true, |expiration| expiration > curr_height)
+        });
+
+        // save the new list
+        self.active.save(store, k.clone(), &active, curr_height)?;
+
+        // load the item
+        let item = self.load_item(store, k, id)?;
+
+        // return the item and the number of items remaining
+        Ok((item, active.len()))
+    }
+
+    /// Updates an item from the vector, returning the new item reference and
+    /// the total number of items remaining. The block height should be greater
+    /// than or equal to the blocks all previous items were added/removed at.
+    /// Updating in the past will lead to incorrect behavior.
+    ///
+    /// This is a convenient way to update items, and reduces the overhead of a
+    /// remove and then push, but it does still create a new item since we need
+    /// to preserve the old item for historical queries. This works even if the
+    /// item is expired since IDs are preserved indefinitely for historical
+    /// queries.
+    pub fn update(
+        &self,
+        store: &mut dyn Storage,
+        k: &K,
+        id: u64,
+        curr_height: u64,
+        update: impl FnOnce(&mut V),
+        expire_in: Option<u64>,
+    ) -> StdResult<(SnapshotVectorMapItemRef, usize)> {
+        // ensure remove operations are performed at or after the last update
+        self.validate_order_and_update(store, k, curr_height)?;
+
+        // get active list for the key
+        let mut active = self.active.may_load(store, k.clone())?.unwrap_or_default();
+
+        // remove existing item and any expired items
+        active.retain(|(active_id, expiration)| {
+            active_id != &id && expiration.map_or(true, |expiration: u64| expiration > curr_height)
+        });
```

Comment:
I feel like by just doing `.retain` on the active list we are opening ourselves up for situations where updating an expired delegation effectively functions as creating a new delegation. Is that the desired outcome here? imho returning an error here would be more clear


```suggestion
        // get active list for the key
        let mut active = self.active.may_load(store, k.clone())?.unwrap_or_default();

        // prior to update, verify that the item exists and is not expired
        let target_id_exists = active.iter().any(|(active_id, expiration)| {
            *active_id == id && expiration.map_or(true, |exp| exp > curr_height)
        });

        if !target_id_exists {
            return Err(StdError::generic_err("item not found or expired"));
        }

        // remove existing item and any expired items
        active.retain(|(active_id, expiration)| {
            active_id != &id && expiration.map_or(true, |expiration: u64| expiration > curr_height)
        });
```

Reply:
I've pushed a test that asserts this: https://github.com/DA0-DA0/dao-contracts/blob/a647cb3f2e8385dfb62a501d45c5e42196b03ddd/packages/cw-snapshot-vector-map/src/tests.rs#L622-L647 

If you feel like this is the expected behavior then please feel free to remove it! I noticed that a few delegation tests fail as a result of this suggested change and I didn't want to alter the core logic before getting your take on it

Reply:
memorializing our conversation here:

yeah, it is the expected behavior that updating an expired item creates a new entry. in fact, updating any item creates a new entry, since we need to preserve old data for historical queries. `update` is just a convenience method for loading a known item by its ID, modifying it, and saving a new item at the current block height.

IMO we don't need to error if the item is expired here because it's not violating any principles of the data structure—the caller should know what they're doing if they're saving an ID and updating it later. if we made this error, and the caller had an ID they wanted to modify, they would have to load it manually, explicitly check it's expiration, conditionally remove it, and then push the modified/new item. this `update` just encapsulates all of that and handles all the cases cleanly. either way, they want to modify some data that they know existed at some point and push a new item, so that's what this `update` is for.